KeepassPinentry/EntryDB.cs

```diff
@@ -3,38 +3,84 @@
 using KeePassLib.Collections;
 using System.Windows.Forms;
 using System;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace KeepassPinentry
 {
+    internal class EntryDBException : Exception
+    {
+        public EntryDBException() { }
+
+        public EntryDBException(string message)
+            : base(message) { }
+    }
+
     internal class EntryDB
     {
         public IPluginHost Host { get; set; }
 
-        public PwEntry Get(string text)
+        public PwEntry GetByTitle(string text)
         {
-            var unlocker = new Unlocker { Host = Host };
-            unlocker.Unlock();
+            return Get(new SearchParameters { SearchInTitles = true, SearchString = text});
+        }
 
-            var p = new SearchParameters
+        public PwEntry GetByUserName(string text)
+        {
+            return Get(new SearchParameters { SearchInUserNames = true, SearchString = text });
+        }
+
+        public PwEntry GetByTitleAndUserName(string title, string name)
+        {
+            var param = new SearchParameters
             {
                 SearchInTitles = true,
-                SearchString = text
+                SearchString = title
             };
 
+            return Get(
+                param,
+                entry =>
+                {
+                    string currentName = entry.Strings.ReadSafe(PwDefs.UserNameField);
+
+                    return string.Equals(currentName, name, StringComparison.OrdinalIgnoreCase);
+                });
+        }
+
+        public PwEntry Get(SearchParameters param, Func<PwEntry, bool> predicate = null)
+        {
+            try
+            {
+                IEnumerable<PwEntry> entries = GetAll(param);
+                PwEntry entry = predicate is null ? entries.First() : entries.First(predicate);
+                NotifyReadEntry(entry);
+
+                return entry;
+            }
+            catch (InvalidOperationException)
+            {
+                var message = $"Unable to find entry with the search string {param.SearchString}.";
+                ShowBalloonNotification(message);
```

Comment:
This message is unfortunately vague since it does not state which fields of the entry are being searched. Furthermore, the predicate can perform some arbitrary check; what it checks for cannot be inferred at this point. I think that for more specific messages, the error would have to be handled in all the higher-level wrappers like `GetByUserName`, but that might feel like duplicate code.